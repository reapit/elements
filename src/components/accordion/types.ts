import { ReactNode, HTMLAttributes } from 'react'

export interface AccordionBaseProps extends HTMLAttributes<HTMLDivElement> {}
export interface AccordionBaseItemProps extends HTMLAttributes<HTMLAnchorElement> {}

export interface AccordionTitleItemProps {}

export interface AccordionItemProps {
  title: string
  titleItems?: ReactNode[]
  content: ReactNode
  onClick?: () => void
}

export interface AccordionProps extends AccordionBaseProps {
  items: AccordionItemProps[]
}
