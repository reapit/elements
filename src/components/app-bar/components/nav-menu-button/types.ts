import { HTMLAttributes } from 'react'
import { MenuItemProps } from '../../../menu/types'
import { IconNames } from '../../../icon'
import { ButtonProps } from '../../../button'

export interface NavMenuButtonContainerBaseProps extends HTMLAttributes<HTMLDivElement> {}
export interface NavMenuButtonBaseProps extends HTMLAttributes<HTMLDivElement> {}

export interface NavMenuTogglerButtonProps extends ButtonProps {
  label: string
  isExpanded: boolean
  onClick: VoidFunction
  icon?: IconNames
}

export interface NavMenuButtonProps extends Pick<NavMenuTogglerButtonProps, 'label' | 'icon'> {
  menuGroups: {
    items: MenuItemProps[]
    title?: string
  }[]
  top?: number
  alignment?: 'left' | 'right'
}

export interface NavMenuButtonMenuProps extends HTMLAttributes<HTMLDivElement> {
  top?: number
}
