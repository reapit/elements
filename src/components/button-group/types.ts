import { HTMLAttributes, ReactNode } from 'react'

export interface ButtonGroupProps extends HTMLAttributes<HTMLDivElement> {
  children: ReactNode
}
