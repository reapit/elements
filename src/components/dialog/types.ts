import { FC, HTMLAttributes } from 'react'

import { ElDialogContainerProps } from './style'

export interface DialogContainerProps extends ElDialogContainerProps {}

export interface DialogTitleProps extends HTMLAttributes<HTMLHeadingElement> {}

export interface DialogBodyProps extends HTMLAttributes<HTMLDivElement> {}

export interface DialogFooterProps extends HTMLAttributes<HTMLDivElement> {}

export type DialogFooterItems = FC<{ closeDialog: VoidFunction }>
