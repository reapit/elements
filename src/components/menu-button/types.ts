import { HTMLAttributes } from 'react'
import { Intent } from '../../helpers/intent'
import { ButtonProps } from '../button'
import { IconNames } from '../icon'
import { MenuItemProps } from '../menu/types'

export interface MenuButtonContainerBaseProps extends HTMLAttributes<HTMLDivElement> {}
export interface MenuButtonBaseProps extends HTMLAttributes<HTMLDivElement> {}

export interface MenuTogglerButtonProps extends ButtonProps {
  label: string
  isExpanded: boolean
  onClick: VoidFunction
  hasBorder?: boolean
  intent?: Intent
  icon?: IconNames
}

export interface MenuButtonProps extends Pick<MenuTogglerButtonProps, 'label' | 'intent' | 'icon' | 'hasBorder'> {
  menuGroups: {
    items: MenuItemProps[]
    title?: string
    intent?: Intent
  }[]
  top?: number
  alignment?: 'left' | 'right'
}

export interface MenuButtonMenuProps extends HTMLAttributes<HTMLDivElement> {
  top?: number
}
