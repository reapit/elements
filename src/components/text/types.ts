// TODO: Ideally we can derive or generate these from the token files in future.
export const fontSizes = ['3xs', '2xs', 'xs', 'sm', 'base', 'md', 'lg', 'xl', '2xl', '3xl'] as const
export const fontWeights = ['regular', 'medium', 'bold'] as const
export const textColours = [
  'primary',
  'secondary',
  'tertiary',
  'white',
  'error',
  'info',
  'success',
  'warning',
  'pending',
  'placeholder',
  'accent_1',
  'accent_2',
  'action',
] as const

export type FontSize = (typeof fontSizes)[number]
export type FontWeight = (typeof fontWeights)[number]
export type TextColour = (typeof textColours)[number]
