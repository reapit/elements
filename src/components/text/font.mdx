import { Meta } from '@storybook/addon-docs/blocks'

<Meta title="Components/Text/Font" />

# `font`

Many components with Elements need to apply font styles based on a specific size and weight. The required styles will
be specified within Figma for each component and will look something like the following CSS:

```css
/* text-sm/Regular */
font-family: var(--font-sm-regular-family, Inter);
font-size: var(--font-sm-regular-size, 14px);
font-style: normal;
font-weight: var(--font-sm-regular-weight, 400);
line-height: var(--font-sm-regular-line_height, 20px); /* 142.857% */
letter-spacing: var(--font-sm-regular-letter_spacing, -0.14px);
```

When an engineer uses this CSS, they will simply copy and paste it into their styled element verbatim, then remove the
CSS variable default/fallback values. The problem with this is we duplicate a lot of font-related CSS throughout many
components, which increases the cost of change, especially if we go through another CSS variable overhaul.

To ease this cost, we have a `font` CSS style helper that can be used. With this helper, instead of duplicating the CSS
from Figma, an engineer will do the following:

```ts
// styles.ts
import { font } from '#src/components/text'

const MyStyledDiv = styled.div`
  ${font('sm', 'regular')}
`
```

This will generate the same CSS as shown in the first snippet, with the exception of `font-style` (generally, we want
to inherit `font-style` rather than force it to be `normal` in all cases).
