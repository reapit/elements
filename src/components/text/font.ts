import type { FontSize, FontWeight } from './types'

/**
 * A helper function for generating a collection of font-related CSS properties for a given font size and
 * weight. The generated string is self-terminating, so there's no need to add a semicolon after interpolating
 * the result into a `css` or `styled` template literal.
 *
 * Importantly, this function intentionally avoids support for text colour. This is because Design System
 * components should have component-specific CSS variables for their text colour. This is not the case for
 * font styles.
 *
 * @param size - The size of the text
 * @param weight - The weight of the text
 * @returns A string of CSS properties for the text
 */
export function font(size: FontSize, weight: FontWeight) {
  const fontFamily = `var(--font-${size}-${weight}-family)`
  const fontSize = `var(--font-${size}-${weight}-size)`
  const fontWeight = `var(--font-${size}-${weight}-weight)`
  const letterSpacing = `var(--font-${size}-${weight}-letter_spacing)`
  const lineHeight = `var(--font-${size}-${weight}-line_height)`

  return `
    /* text-${size}/${weight} */
    font: ${fontWeight} ${fontSize}/${lineHeight} ${fontFamily};
    letter-spacing: ${letterSpacing};
  `
}
