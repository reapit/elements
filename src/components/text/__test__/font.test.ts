import { test, expect } from 'vitest'
import { font } from '../font'
import { fontSizes, fontWeights } from '../types'

const testCases = fontSizes.flatMap((size) => fontWeights.map((weight) => [size, weight] as const))

test.each(testCases)('font(%s, %s) returns correct CSS', (size, weight) => {
  expect(font(size, weight)).toMatchSnapshot()
})
