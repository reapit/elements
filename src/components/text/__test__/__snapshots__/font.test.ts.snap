// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`font(2xl, bold) returns correct CSS 1`] = `
"
    /* text-2xl/bold */
    font: var(--font-2xl-bold-weight) var(--font-2xl-bold-size)/var(--font-2xl-bold-line_height) var(--font-2xl-bold-family);
    letter-spacing: var(--font-2xl-bold-letter_spacing);
  "
`;

exports[`font(2xl, medium) returns correct CSS 1`] = `
"
    /* text-2xl/medium */
    font: var(--font-2xl-medium-weight) var(--font-2xl-medium-size)/var(--font-2xl-medium-line_height) var(--font-2xl-medium-family);
    letter-spacing: var(--font-2xl-medium-letter_spacing);
  "
`;

exports[`font(2xl, regular) returns correct CSS 1`] = `
"
    /* text-2xl/regular */
    font: var(--font-2xl-regular-weight) var(--font-2xl-regular-size)/var(--font-2xl-regular-line_height) var(--font-2xl-regular-family);
    letter-spacing: var(--font-2xl-regular-letter_spacing);
  "
`;

exports[`font(2xs, bold) returns correct CSS 1`] = `
"
    /* text-2xs/bold */
    font: var(--font-2xs-bold-weight) var(--font-2xs-bold-size)/var(--font-2xs-bold-line_height) var(--font-2xs-bold-family);
    letter-spacing: var(--font-2xs-bold-letter_spacing);
  "
`;

exports[`font(2xs, medium) returns correct CSS 1`] = `
"
    /* text-2xs/medium */
    font: var(--font-2xs-medium-weight) var(--font-2xs-medium-size)/var(--font-2xs-medium-line_height) var(--font-2xs-medium-family);
    letter-spacing: var(--font-2xs-medium-letter_spacing);
  "
`;

exports[`font(2xs, regular) returns correct CSS 1`] = `
"
    /* text-2xs/regular */
    font: var(--font-2xs-regular-weight) var(--font-2xs-regular-size)/var(--font-2xs-regular-line_height) var(--font-2xs-regular-family);
    letter-spacing: var(--font-2xs-regular-letter_spacing);
  "
`;

exports[`font(3xl, bold) returns correct CSS 1`] = `
"
    /* text-3xl/bold */
    font: var(--font-3xl-bold-weight) var(--font-3xl-bold-size)/var(--font-3xl-bold-line_height) var(--font-3xl-bold-family);
    letter-spacing: var(--font-3xl-bold-letter_spacing);
  "
`;

exports[`font(3xl, medium) returns correct CSS 1`] = `
"
    /* text-3xl/medium */
    font: var(--font-3xl-medium-weight) var(--font-3xl-medium-size)/var(--font-3xl-medium-line_height) var(--font-3xl-medium-family);
    letter-spacing: var(--font-3xl-medium-letter_spacing);
  "
`;

exports[`font(3xl, regular) returns correct CSS 1`] = `
"
    /* text-3xl/regular */
    font: var(--font-3xl-regular-weight) var(--font-3xl-regular-size)/var(--font-3xl-regular-line_height) var(--font-3xl-regular-family);
    letter-spacing: var(--font-3xl-regular-letter_spacing);
  "
`;

exports[`font(3xs, bold) returns correct CSS 1`] = `
"
    /* text-3xs/bold */
    font: var(--font-3xs-bold-weight) var(--font-3xs-bold-size)/var(--font-3xs-bold-line_height) var(--font-3xs-bold-family);
    letter-spacing: var(--font-3xs-bold-letter_spacing);
  "
`;

exports[`font(3xs, medium) returns correct CSS 1`] = `
"
    /* text-3xs/medium */
    font: var(--font-3xs-medium-weight) var(--font-3xs-medium-size)/var(--font-3xs-medium-line_height) var(--font-3xs-medium-family);
    letter-spacing: var(--font-3xs-medium-letter_spacing);
  "
`;

exports[`font(3xs, regular) returns correct CSS 1`] = `
"
    /* text-3xs/regular */
    font: var(--font-3xs-regular-weight) var(--font-3xs-regular-size)/var(--font-3xs-regular-line_height) var(--font-3xs-regular-family);
    letter-spacing: var(--font-3xs-regular-letter_spacing);
  "
`;

exports[`font(base, bold) returns correct CSS 1`] = `
"
    /* text-base/bold */
    font: var(--font-base-bold-weight) var(--font-base-bold-size)/var(--font-base-bold-line_height) var(--font-base-bold-family);
    letter-spacing: var(--font-base-bold-letter_spacing);
  "
`;

exports[`font(base, medium) returns correct CSS 1`] = `
"
    /* text-base/medium */
    font: var(--font-base-medium-weight) var(--font-base-medium-size)/var(--font-base-medium-line_height) var(--font-base-medium-family);
    letter-spacing: var(--font-base-medium-letter_spacing);
  "
`;

exports[`font(base, regular) returns correct CSS 1`] = `
"
    /* text-base/regular */
    font: var(--font-base-regular-weight) var(--font-base-regular-size)/var(--font-base-regular-line_height) var(--font-base-regular-family);
    letter-spacing: var(--font-base-regular-letter_spacing);
  "
`;

exports[`font(lg, bold) returns correct CSS 1`] = `
"
    /* text-lg/bold */
    font: var(--font-lg-bold-weight) var(--font-lg-bold-size)/var(--font-lg-bold-line_height) var(--font-lg-bold-family);
    letter-spacing: var(--font-lg-bold-letter_spacing);
  "
`;

exports[`font(lg, medium) returns correct CSS 1`] = `
"
    /* text-lg/medium */
    font: var(--font-lg-medium-weight) var(--font-lg-medium-size)/var(--font-lg-medium-line_height) var(--font-lg-medium-family);
    letter-spacing: var(--font-lg-medium-letter_spacing);
  "
`;

exports[`font(lg, regular) returns correct CSS 1`] = `
"
    /* text-lg/regular */
    font: var(--font-lg-regular-weight) var(--font-lg-regular-size)/var(--font-lg-regular-line_height) var(--font-lg-regular-family);
    letter-spacing: var(--font-lg-regular-letter_spacing);
  "
`;

exports[`font(md, bold) returns correct CSS 1`] = `
"
    /* text-md/bold */
    font: var(--font-md-bold-weight) var(--font-md-bold-size)/var(--font-md-bold-line_height) var(--font-md-bold-family);
    letter-spacing: var(--font-md-bold-letter_spacing);
  "
`;

exports[`font(md, medium) returns correct CSS 1`] = `
"
    /* text-md/medium */
    font: var(--font-md-medium-weight) var(--font-md-medium-size)/var(--font-md-medium-line_height) var(--font-md-medium-family);
    letter-spacing: var(--font-md-medium-letter_spacing);
  "
`;

exports[`font(md, regular) returns correct CSS 1`] = `
"
    /* text-md/regular */
    font: var(--font-md-regular-weight) var(--font-md-regular-size)/var(--font-md-regular-line_height) var(--font-md-regular-family);
    letter-spacing: var(--font-md-regular-letter_spacing);
  "
`;

exports[`font(sm, bold) returns correct CSS 1`] = `
"
    /* text-sm/bold */
    font: var(--font-sm-bold-weight) var(--font-sm-bold-size)/var(--font-sm-bold-line_height) var(--font-sm-bold-family);
    letter-spacing: var(--font-sm-bold-letter_spacing);
  "
`;

exports[`font(sm, medium) returns correct CSS 1`] = `
"
    /* text-sm/medium */
    font: var(--font-sm-medium-weight) var(--font-sm-medium-size)/var(--font-sm-medium-line_height) var(--font-sm-medium-family);
    letter-spacing: var(--font-sm-medium-letter_spacing);
  "
`;

exports[`font(sm, regular) returns correct CSS 1`] = `
"
    /* text-sm/regular */
    font: var(--font-sm-regular-weight) var(--font-sm-regular-size)/var(--font-sm-regular-line_height) var(--font-sm-regular-family);
    letter-spacing: var(--font-sm-regular-letter_spacing);
  "
`;

exports[`font(xl, bold) returns correct CSS 1`] = `
"
    /* text-xl/bold */
    font: var(--font-xl-bold-weight) var(--font-xl-bold-size)/var(--font-xl-bold-line_height) var(--font-xl-bold-family);
    letter-spacing: var(--font-xl-bold-letter_spacing);
  "
`;

exports[`font(xl, medium) returns correct CSS 1`] = `
"
    /* text-xl/medium */
    font: var(--font-xl-medium-weight) var(--font-xl-medium-size)/var(--font-xl-medium-line_height) var(--font-xl-medium-family);
    letter-spacing: var(--font-xl-medium-letter_spacing);
  "
`;

exports[`font(xl, regular) returns correct CSS 1`] = `
"
    /* text-xl/regular */
    font: var(--font-xl-regular-weight) var(--font-xl-regular-size)/var(--font-xl-regular-line_height) var(--font-xl-regular-family);
    letter-spacing: var(--font-xl-regular-letter_spacing);
  "
`;

exports[`font(xs, bold) returns correct CSS 1`] = `
"
    /* text-xs/bold */
    font: var(--font-xs-bold-weight) var(--font-xs-bold-size)/var(--font-xs-bold-line_height) var(--font-xs-bold-family);
    letter-spacing: var(--font-xs-bold-letter_spacing);
  "
`;

exports[`font(xs, medium) returns correct CSS 1`] = `
"
    /* text-xs/medium */
    font: var(--font-xs-medium-weight) var(--font-xs-medium-size)/var(--font-xs-medium-line_height) var(--font-xs-medium-family);
    letter-spacing: var(--font-xs-medium-letter_spacing);
  "
`;

exports[`font(xs, regular) returns correct CSS 1`] = `
"
    /* text-xs/regular */
    font: var(--font-xs-regular-weight) var(--font-xs-regular-size)/var(--font-xs-regular-line_height) var(--font-xs-regular-family);
    letter-spacing: var(--font-xs-regular-letter_spacing);
  "
`;
